### PR TITLE
feat(auth): Add Expo support for phone auth

### PR DESCRIPTION
**Pull request by nandorojo**
- Add `@react-native-firebase/auth` config plugin, which enables captcha for phone auth by modifying `Info.Plist`. 
  - The config plugin is automating the [reCAPTCHA verification steps](https://firebase.google.com/docs/auth/ios/phone-auth#set-up-recaptcha-verification) from the firebase iOS docs.
- Add docs to the **Phone Auth** page for Expo users.

Sorry for some reason the checklist didn't show up here when I created the PR. I'll try to find it.

Related: #5936

### CODE REVIEW

File: packages/auth/plugin/src/ios/urlTypes.ts
Comment (barthap) [inline packages/auth/plugin/src/ios/urlTypes.ts:13]: In theory, we don't have to use `withInfoPlist`, because it is an [asynchronous mod]() and since here we don't have any async code, maybe we can keep it simple: 

https://github.com/expo/expo-cli/blob/cdb715365bab7f735200c32971437de187a9953d/packages/config-plugins/src/plugins/ios-plugins.ts#L79

File: packages/auth/plugin/src/ios/urlTypes.ts
Comment (barthap) [inline packages/auth/plugin/src/ios/urlTypes.ts:24]: I'd add unit tests for this method. To keep it more simple, I'd split it into two functions:
- `getReservedClientId(googleServicesFilePath: string): string`
- `addUriScheme(config, scheme): config`
 
Then provide fixtures for `GoogleServices-Info.plist` and some simple config object `const config = { ios: { infoPlist: { ... } }` and compare outputs.

I can take care of that in the subsequent PR if you wish.

Comment (barthap) [inline packages/auth/plugin/src/ios/urlTypes.ts:24]: I'd add unit tests for this method. To keep it more simple, I'd split it into two functions:
- `getReservedClientId(googleServicesFilePath: string): string`
- `addUriScheme(config, scheme): config`
Then provide fixtures for `GoogleServices-Info.plist` and some simple config object `const config = { ios: { infoPlist: { ... } }` and compare outputs.

**Review by barthap — CHANGES_REQUESTED** [review]
The plugin code itself looks good 👏  I only left a suggestion about the possibility of adding unit tests.

However, for the plugin to be working with the `@rnfb/auth` package, two things are missing:
- the `app.plugin.js` file in the `auth` package directory: https://github.com/invertase/react-native-firebase/blob/main/packages/crashlytics/app.plugin.js
- `package.json` scripts for building the plugin TS code: https://github.com/invertase/react-native-firebase/blob/ef7895cd49eed3a4d103df624e52315250c68ab3/packages/crashlytics/package.json#L11-L13

**Comment by mikehardy** [comment]
@nandorojo this appears to be close, but with some nice + detailed comments still needing handling that should get it to the finish line. I marked it as waiting for response

File: packages/auth/plugin/src/ios/urlTypes.ts
Comment (nandorojo) [reply packages/auth/plugin/src/ios/urlTypes.ts:24]: Would be great if you wouldn't mind writing the unit test. Unit tests aren't something I'm that familiar with. Appreciate the offer to help.

Comment (nandorojo) [reply packages/auth/plugin/src/ios/urlTypes.ts:24]: As for the functions, I should be able to split it up.

**Comment by nandorojo** [comment]
@barthap thanks for your comments here! I'll try to address them shortly. Would be great if you could add the aforementioned unit test too once this is merged.

File: packages/auth/plugin/src/ios/urlTypes.ts
Comment (nandorojo) [inline packages/auth/plugin/src/ios/urlTypes.ts:57]: this should be `if (!`, I think I messed that up.

File: packages/auth/plugin/src/ios/urlTypes.ts
Comment (barthap) [reply packages/auth/plugin/src/ios/urlTypes.ts:24]: Ok, so go on with this PR, I'd add some tests then 😉

**Comment by justin-chau** [comment]
Are there any updates on this pull request?

**Comment by mikehardy** [comment]
None that I've seen, I don't think there is any harm in taking the changes on the fork and proposing a PR based on it if this is functionality you need

**Comment by nandorojo** [comment]
Sorry I’ve been too busy to get to this. @justin-chau you’re welcome to apply the suggested edits as commits to my PR and I can accept them

**Comment by justin-chau** [comment]
Is the only thing missing at the moment the unit tests?

**Comment by barthap** [comment]
My review comments are still unaddressed, especially the one about missing `package.json` scripts and missing `app.plugin.js` file.

**Comment by nandorojo** [comment]
I addressed all the feedback. @barthap let me know if this is good to go.

File: packages/auth/package.json
Comment (barthap) [inline packages/auth/package.json:28]:
```suggestion
    "@expo/config-plugins": "^4.1.4"
```
Other packages have this version as dependency

File: packages/auth/package.json
Comment (barthap) [reply packages/auth/package.json]: I have no opinion on this, maybe @mikehardy would tell more 😉

**Review by barthap — DISMISSED** [review]
Should be good now 👏  I'll add unit tests in a separate PR soon

File: packages/auth/package.json
Comment (nandorojo) [reply packages/auth/package.json]: wouldn it be best for the plug-in to have its own node modules?

Comment (mikehardy) [reply packages/auth/package.json]: No - over long periods of time, maintaining dependencies (and forward-porting across breaking changes) is one of the most time-consuming parts of maintaining a repo, and the effort requires scales with the number packages.json's / node_modules - so I want fewer if possible

**Comment by nandorojo** [comment]
Just need to run `yarn` to resolve the lock file conflict and this will be good to go. I’ll do it when I’m at my computer.

**Comment by izakfilmalter** [comment]
Any updates on this? Would love to get this merged.

**Comment by mikehardy** [comment]
I rebased + re-pushed this - I need to re-scan it prior to merge but last time I did (quite a time ago!) it seemed fine

Unfortunately in the time between proposal and now firebase-ios-sdk v9 was adopted (in react-native-firebase v15) which requires `use_frameworks`, and I just mention that because I know that this transition is still in process in a lot of ways and is difficult to adopt. The related issues here are still open with users helping each other figure out exactly how best to get it working for Expo, if I understand correctly - and even in non-Expo case you have to disable Flipper + Hermes for it to work on iOS

**Comment by mikehardy** [comment]
:thinking: Hmm some sort of entropy appears to have taken over, I probably will not have time to fix this for a while (weeks perhaps) as I'm traveling at the moment

```
Run node -e "require('./website/scripts/generate-typedoc').generateTypedoc()"
/home/runner/work/react-native-firebase/react-native-firebase/tsconfig.json
Using TypeScript 3.9.10 from /home/runner/work/react-native-firebase/react-native-firebase/website/node_modules/typescript/lib
Error: /home/runner/work/react-native-firebase/react-native-firebase/node_modules/@types/jest/index.d.ts([4](https://github.com/invertase/react-native-firebase/runs/7396232937?check_suite_focus=true#step:9:5)1[5](https://github.com/invertase/react-native-firebase/runs/7396232937?check_suite_focus=true#step:9:6))
 ']' expected.
Error: /home/runner/work/react-native-firebase/react-native-firebase/node_modules/@types/jest/index.d.ts(415)
 ';' expected.
Error: /home/runner/work/react-native-firebase/react-native-firebase/node_modules/@types/jest/index.d.ts(415)
 ';' expected.
Error: /home/runner/work/react-native-firebase/react-native-firebase/node_modules/@types/jest/index.d.ts(415)
 ';' expected.
Error: /home/runner/work/react-native-firebase/react-native-firebase/node_modules/@types/jest/index.d.ts(415)
 Declaration or statement expected.
Error: /home/runner/work/react-native-firebase/react-native-firebase/node_modules/@types/jest/index.d.ts(419)
 ']' expected.
Error: /home/runner/work/react-native-firebase/react-native-firebase/node_modules/@types/jest/index.d.ts(419)
 ';' expected.
Error: /home/runner/work/react-native-firebase/react-native-firebase/node_modules/@types/jest/index.d.ts(419)
 ';' expected.
Error: /home/runner/work/react-native-firebase/react-native-firebase/node_modules/@types/jest/index.d.ts(419)
 ';' expected.
Error: /home/runner/work/react-native-firebase/react-native-firebase/node_modules/@types/jest/index.d.ts(419)
 Declaration or statement expected.
Error: /home/runner/work/react-native-firebase/react-native-firebase/node_modules/@types/jest/index.d.ts(419)
 Declaration or statement expected.
Error: /home/runner/work/react-native-firebase/react-native-firebase/node_modules/@types/jest/index.d.ts(13[8](https://github.com/invertase/react-native-firebase/runs/7396232937?check_suite_focus=true#step:9:9)4)
 Declaration or statement expected.
node:internal/fs/utils:345
    throw err;
    ^
```

I welcome any efforts here

**Comment by nandorojo** [comment]
Does that only happen on this branch? What causes that to happen?

**Comment by mikehardy** [comment]
> Does that only happen on this branch? What script causes that to happen?

It's in CI, so you can see all related dependencies + commands run by looking at the related github workflows and github actions log of the run

It appears to be only this branch, but I just rebased it against main (which is working) and I don't know why this branch would cause the problem - I have not investigated at all. I will re-run the action in case it was just some transient flaky failure

**Comment by mikehardy** [comment]
Vercel deployment failed for the same reason. Seems strange, I don't know sorry :shrug: - I'm about to be away from keyboard and will likely not have time to inspect for weeks, I'm not sure. Worst case, could effectively create a new branch from main then re-apply the changes since they are probably self-contained ? To make sure it's "clean"

**Comment by pduggi-cf** [comment]
I rebased and pushed using the latest main, my PR is #6472 
Please review it and merge it if it makes more sense

**Comment by mikehardy** [comment]
Superceded by #6645

**Pull request closed without merging**
